### DISCUSSION TITLE
Can I supply the private key used for secret encryption via environment?

### DISCUSSION THREAD

**Discussion by konstantinblaesi** [renovatebot/renovate#10182] [Request Help]
Using the documentations openssl commands I generated public/private RSA keys and encrypted my configuration.

Private key looks as usual:

> -----BEGIN RSA PRIVATE KEY-----
> ...
> ...
> ...
> -----END RSA PRIVATE KEY-----
> 

I am running renovate using the image renovate/renovate. How can I supply the private key, does it have to be via file? I tried via environment variable RENOVATE_PRIVATE without luck and I am wondering whether this is supported at all. How can/should it be done ?

**Comment by rarkins** [comment]
`privateKey` in config maps to `RENOVATE_PRIVATE_KEY` in env.

**Reply by konstantinblaesi** [reply]
Ah yes I had a typo, I was already using RENOVATE_PRIVATE_KEY . I think my problem is how to properly map the newlines of the private key file to an environment variable. The NodeJS crypto.privateDecrypt(...) is not happy with the structure of the key :(

**Reply by viceice** [reply]
Renovate needs the private key as `privateKey` and you need to use the public key to encrypt the values.

I think you need to  simply copy the not protected private key to gitlab `RENOVATE_PRIVATE_KEY` variable with all linebreaks as is.
So make sure the private key isn't password protected.

**Reply by konstantinblaesi** [reply]
Correct, seems I am catching all the typos today. I have generated a keypair with the openssl commands as documented, I have written a small NodeJS script that parses my renovate config and encrypts password or token values so that
```
{
  ...
  "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "some-team.azurecr.io",
      "username": "some-user",
      "password": "some-password"
    }
  ]
}
```
becomes
```
{
  ...
  "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "some-team.azurecr.io",
      "username": "some-user",
      "encrypted": {
        "password": "encrypted-value-here"
      }
    }
  ]
}
```

Snippet for encrypting:
```
crypto.publicEncrypt({
        key: publicKey,
        padding: crypto.constants.RSA_PKCS1_PADDING
    }, Buffer.from(valueToEncrypt, 'utf-8'))
        .toString("base64");
```

Snippet for decrypting:
```
crypto
    .privateDecrypt(
        {
            key: privateKey,
            padding: crypto.constants.RSA_PKCS1_PADDING,
        },
        Buffer.from(valueToDecrypt, 'base64')
    )
    .toString();
```

My snippet has no issues decryting encrypted values if I read the private key from the file or assign the private key directly like so:
```
privateKey = "-----BEGIN RSA PRIVATE KEY-----\n" +
    "..............\n" +
    // etc.
    "-----END RSA PRIVATE KEY-----\n";
```

If I use RENOVATE_PRIVATE_KEY including `\n` or `\\n` I just get an error from the renovate docker container
> Failed to decrypt field password. Please re-encrypt and try again.

It seems like my issues are how [environment variables, newlines and node interact](https://stackoverflow.com/questions/30400341/environment-variables-containing-newlines-in-node/45889979)?

**Reply by viceice** [reply]
can you test in node if `process.env.RENOVATE_PRIVATE_KEY === privateKey` ?

Do not try to do this locally:

```bash
export X="hey\nman\ndude\nwhat"
node -e 'console.log(process.env.X)'
```

as the shell string is already escaped, so you get the newlines as `\\n`, you really need to make sure your gitlab variable has the newlines and you don't manipulate it before passing to renovate.


better try this locally:

```bash
export X=$(cat keyfile)
node -e 'console.log(process.env.X)'
```

This should preserve newlines in env
